neuralpp/experiments/experimental_inference/graph_computation.py

```diff
@@ -1,88 +1,74 @@
         self.tree = tree
         self.result_dict = {}
 
+    def __contains__(self, item):
+        return id(item) in self.result_dict
+
     def __getitem__(self, item):
+        if item in self:
+            return self.result_dict[id(item)]
+        self.compute_result_dict(item)
         return self.result_dict[id(item)]
 
     def __setitem__(self, item, value):
         self.result_dict[id(item)] = value
 
     def __call__(self, item):
-        self.compute_result_dict(item)
-        return self.result_dict[id(item)]
+        return self[item]
 
     def compute_result_dict(self, node):
-        raise NotImplementedError()
-
-    def update_value(self, target_node, child_value):
-        raise NotImplementedError()
-
-    def recompute_and_propagate_result_to_ancestors(self, node):
-        self.compute_result_dict(node)
-        node_result = self[node]
-        target_node = self.tree.parent(node)
-        while target_node is not None:
-            self.update_value(target_node, node_result)
-            target_node = self.tree.parent(target_node)
-
-
-Expansion = namedtuple("Expansion", "node expansion_value")
-
-
-class ExpansionValueComputation(TreeComputation):
-
-    def __init__(self, partial_tree, full_tree, expansion_value_function):
-        super().__init__(partial_tree)
-        self.full_tree = full_tree
-        self.expansion_value_function = expansion_value_function
-        self.compute_result_dict(partial_tree.root)
-
-    def __compute_single_node(self, node, get_child_scores_from_cache: bool):
-        children = self.full_tree.children(node)
-        if not get_child_scores_from_cache:
-            for child in children:
-                self.compute_result_dict(child)
-        candidates = [self[child] for child in children if self[child] is not None]
-        return argmax(candidates, lambda node_value_pair: node_value_pair.expansion_value)
-
-    def compute_result_dict(self, node):
-        if node not in self.tree:
-            self[node] = Expansion(node, self.expansion_value_function(node, self.tree, self.full_tree))
-        else:
-            self[node] = self.__compute_single_node(node, False)
-
-    def update_value(self, target_node, child_value):
-        ancestor_value = self.result_dict.get(id(target_node))
-        assert(ancestor_value is not None)
-        if child_value is None or child_value.expansion_value <= ancestor_value.expansion_value:
-            self[target_node] = self.__compute_single_node(target_node, True)
+        self[node] = self.compute(node)
+
+    def invalidate(self, node):
+        """
+        Invalidate cached value for `node` and its ancestors.
+        """
+        while node is not None:
+            if node in self:
+                del self.result_dict[id(node)]
+            node = self.tree.parent(node)
+
+
+class MaximumLeafValueComputation(TreeComputation):
+
+    def __init__(self, tree: Tree, leaf_value_function, argmax_key=None):
```

Comment:
Is `argmax_key` really optional? It seems there will be an error if it is `None`.

Reply:
Yes, as of 3.8: https://docs.python.org/3/library/functions.html#max

As far as I can tell it goes by identity/usual max function if the key is None.